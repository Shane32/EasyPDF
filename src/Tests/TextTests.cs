using System.Drawing;
using System.Text;
using Font = Shane32.EasyPDF.Font;

namespace Tests;

public class TextTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public TextTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Fact]
    public void PositioningTests()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
