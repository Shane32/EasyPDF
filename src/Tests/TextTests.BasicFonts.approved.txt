%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 914/Filter/FlateDecode>>stream
x\x9C}\xD9Mo\xD40\x10\xC6\xF1{>E\x8Ep\xC0\xC4\x8E\x9D\xC4W$^\xC4\x94\x1B\xE2\x04mE\xB5\x05\xB5 \xF1\xF5I\xE6\x19\x01R\xFFSA\xCB\x92_\xBB;~f\xE2d\xB5\xF7\xC3\xAB}\x98\x97qm\xCB\xB8\x7F\x1D^\xEF\xC3\x87!\x8F\xD3\xF1\xE7E\x1E\xD7r\xFC\x9D\xC6/w\xC34\xBE?\xBEn\x87)\xE5\xF1\xF7\xF0\xE9\xF3\xE1_\x87\xC9~\xEE\xE3[\x7F\xF0p\xC3?v\x7F\xBE\xC4\xCB7y\xCCe\xDC\xAF\xFF=\xFB\xF9\xCF~\xF7\xFF/Og\x09\xCF\xDE]].?\x9E\xEF\xB7*fK\xEB<\xE6\x9E\xB6y,)o\xFA\xF6p5\x1CO4\xA5\x9E\x03\x9A\xD3\xD4\x03j\xA9\xAC\x01\xAD\xA9\xB6\x80zZ*S\xC9i+L\xF3\x94Z\xF0\x84s\x8B\x9F{Xa\xCDa\x85u\x0E+\xAC-\xAC\xB0n|\xBCM\xE7C\xA6\x92\x8EyA\xB2^\x95\x92\xA6\xCC!!)\x09&K\x02IU \xA9\x8A\x9Ar\xE7\xB1`\xB2\xB1`\xB2\xB1@\xF2e!-\xF6\x7F"\x8D\x05S\x0F\xCBP\x83\x99ZXFj\xF0\xFC\x88\x94\xDF\x92\xE65\xC8\x0FI\xF9!)?"\xCF\x8F\xC8\xC7\x02\xA9\x87\xAF\xE5!!\xB5\xF0\xB5\xCE\x90\xE8\xB8\x87D\xA4\x906k$\x86\x84\xA4\x90\x90\x14\x12\x91\x87\x84\xB4Y\x81Ds\x09+\xF4h\x91zX\xA1G\x8B\xD4\xC2
\xCFh\xE9\xB8GKd\xD1\x1E\xE7\xC8Z9\x09&K\x02I\xA7\x1BS\x09_KI \xA9v$\xD5^R/|\xB5b\xB2\x89a\xB2\x89a\xB2\x89a\xB2\xAB\x15\x92G\x88d;\x16\x92G\x88T\xC3uy\x84D\xBA\x901\xCDa\xF1\x9A3\xA6\xA0r]\xC8\x98\xD4\xC6'\xA2h\xE9\xF1%II\x80\xA8\xF5k*[\xD0z$\xB5\x1EI\xFDER\x7F\x89\xBC\xBFH\xD1\xAFh\xB1D\xBEZ\xA4\x1A.\xEA8\x18-\xCAG\x82\xA8\x065(\xD7\x9E\xEA\x12\xE4\x8A\xA4\x84\x88Jt\\1\x10\xF96\x81\xA4\xB5\x12\xF9\x8C\x13\xF9T"i*\xE3\xF5\x1EO\xBB\x04\x17\x1D&\xDBB\x98,%$\xCD\x11S\xF4+\x16 \x92\x02dZ\xC3\xCA5,H\xCA\x96\xA9\x85\x95\xD7\xA0<\xB5\x83\xC9\xDA\xF1T\xE63\xBD'\xA8\xC1q\xBBJ!\xE9TcRzHv\x162\xADay\x9E\x1E\x92\xED\xBEH\x9E\x12\x91\x9D\xA3u\x81\x9Bq\x0F\x09\xC968$\x9F>\xA4\xCA\xC7=$\xA45\xAC\xC1G\x8C\xC8CBjay\x1E\x12\x92F\x09\x17\xA5Sg\x0B\xDF,0\xA9x"\xAF\x90\xC8zu\x94\x097\xEA\xB6\xBF0Y\x1B\x99l\x7Fa\xB2\x0E3\xD9\xD6\x83\xA4\xE63YNH:\x13\x904\xD3L-|-u\x92\xC9:\x89\xA4tKx\x1B\xCF\xA4e!\xD9\x8E\x81\xA4\xF6#y\xEDD*\xB0\xC2\xBD\xAB\xF7\x18I=FR\x8F\x89|\xC5HZ1\x91\x06\x9EI=F\xEAa\x85\xDE~\xA4\x16V\xE8\x11"\xA9\xFDDJw\x09o\xF2\x99\x94.\x92\xD2%\xF2t\x9144D\x9E.R\x09\x8B\xD7\x0E\x8A\xA4m\x12\xC9#$RN\xFD\xF1]\xB4\xC7D\xA2\x94H\x14\x12\x88gD\xA2\x01$Qx \x1E\x10\xC9\x1AU\xEDs\x09\xE2\xC1\x91pa>v \x96\xE6\x92\xC3\xB7\x03L\x96\x00\x93E\x80\xA4\xF9a*a\x19:q\x99,:$%\xC4d\xE74S\x0B\x97\xAC\x81DR\x84sx\xD7\xCDd\xC3\xCAd\xD3\xCAd\xE3\xCAdWE$\xEF$\x92:\x89\xA4N\x12y'\x91J\x98\x86\xB7\x0B\xA9\x87K\xF6N"\x05\x95{\xAF\x90J\xB8\xA8\xF8\xB3\x86\xB5\x9EWk\xFA\xBCa\xBF\xFA\xF9\xEB\xDB\xF7\x9B\xBF\x9F8\xD8\xA7\x1E\xE3\xDD\x90k=\x1F\\\xEC\xC1\xF9u9I\x0F\xAE\x87?\xE2g\x08\xBB
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001153 00000 n 
0000000015 00000 n 
0000001241 00000 n 
0000000996 00000 n 
0000001306 00000 n 
0000001351 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 